flink-runtime/src/main/java/org/apache/flink/runtime/metrics/ReporterSetup.java

```diff
@@ -464,6 +466,21 @@ private static Optional<MetricReporter> loadViaReflection(
                     alternativeFactoryClassName, reporterName, reporterConfig, reporterFactories);
         }
 
-        return Optional.of((MetricReporter) reporterClass.newInstance());
+        try {
+            return Optional.of((MetricReporter) reporterClass.newInstance());
+        } catch (InstantiationException e) {
+            if (ExceptionUtils.findThrowable(e, NoSuchMethodException.class).isPresent()) {
```

Comment:
This may not be exhaustive enough. Reporters that are exclusively factory-based may not even be public anymore.